includes/class-indieauth-token-endpoint.php

```diff
@@ -0,0 +1,221 @@
+<?php
+/**
+ *
+ *
+ * Implements IndieAuth Token Endpoint
+ */
+
+class IndieAuth_Token_Endpoint {
+	public function __construct() {
+		add_action( 'rest_api_init', array( $this, 'register_routes' ) );
+
+	}
+
+
+		/**
+		 * Extracts the token from the given authorization header.
+		 *
+		 * @param string $header Authorization header.
+		 *
+		 * @return string|null Token on success, null on failure.
+		 */
+	public function get_token_from_bearer_header( $header ) {
+		if ( is_string( $header ) && preg_match( '/Bearer ([\x20-\x7E]+)/', trim( $header ), $matches ) ) {
+				return $matches[1];
+		}
+			return null;
+	}
+
+	/**
+	 * Register the Route.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			'indieauth/1.0', '/token', array(
+				array(
+					'methods'  => WP_REST_Server::CREATABLE,
+					'callback' => array( $this, 'post' ),
+					'args'     => array(
+						'grant_type'   => array(),
+						'code'         => array(),
+						'client_id'    => array(
+							'validate_callback' => array( $this, 'is_valid_url' ),
+							'sanitize_callback' => 'esc_url_raw',
+						),
+						'redirect_uri' => array(
+							'validate_callback' => array( $this, 'is_valid_url' ),
+							'sanitize_callback' => 'esc_url_raw',
+						),
+						'me'           => array(
+							'validate_callback' => array( $this, 'is_valid_url' ),
+							'sanitize_callback' => 'esc_url_raw',
+						),
+						'action'       => array(),
+						'token'        => array(),
+					),
+				),
+			)
+		);
+		register_rest_route(
+			'indieauth/1.0', '/token', array(
+				array(
+					'methods'  => WP_REST_Server::READABLE,
+					'callback' => array( $this, 'get' ),
+					'args'     => array(),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Returns if valid URL for REST validation
+	 *
+	 * @param string $url
+	 *
+	 * @return boolean
+	 */
+	public static function is_valid_url( $url, $request = null, $key = null ) {
+		if ( ! is_string( $url ) || empty( $url ) ) {
+			return false;
+		}
+		return filter_var( $url, FILTER_VALIDATE_URL );
+	}
+
+	public function generate() {
+		return wp_generate_password( 22, false );
```

Comment:
Might as well use a longer string here to be safe. 128 characters isn't so long that it will cause a problem with any clients. And since you'll be storing a hash of the token anyway, it won't increase your storage requirements.